src/Libraries/DSCPython/DCPythonCodeCompletionProviderCore.cs

```diff
@@ -0,0 +1,394 @@
+        /// <returns>The type as an object</returns>
+        private object LookupMember(string name)
+        {
+            var periodIndex = name.IndexOf('.');
+            using (Py.GIL())
+            {
+                if (periodIndex == -1)
+                {
+                    if (Scope.TryGet(name, out object varOutput))
+                    {
+                        return varOutput;
+                    }
+                }
+                return null;
+            }
+        }
+        #endregion
+
+        #region PythonCodeCompletionProviderCommon public methods implementations
+
+        /// <summary>
+        /// Generate completion data for the specified text, while import the given types into the
+        /// scope and discovering variable assignments.
+        /// </summary>
+        /// <param name="code">The code to parse</param>
+        /// <param name="expand">Determines if the entire namespace should be used</param>
+        /// <returns>Return a list of IExternalCodeCompletionData </returns>
+        public override IExternalCodeCompletionData[] GetCompletionData(string code, bool expand = false)
+        {
+            IEnumerable<PythonCodeCompletionDataCore> items = new List<PythonCodeCompletionDataCore>();
+
+            if (!PythonEngine.IsInitialized)
+            {
+                PythonEngine.Initialize();
+                PythonEngine.BeginAllowThreads();
+            }
+
+            try
+            {
+                using (Py.GIL())
+                {
+                    if (code.Contains("\"\"\""))
+                    {
+                        code = StripDocStrings(code);
+                    }
+
+                    if (Scope == null)
+                    {
+                        Scope = CreateUniquePythonScope();
+                    }
+
+                    UpdateImportedTypes(code);
+                    UpdateVariableTypes(code);  // Possibly use hindley-milner in the future
+
+                    // If expand param is true use the entire namespace from the line of code
+                    // Else just return the last name of the namespace
+                    string name = expand ? GetLastNameSpace(code) : GetLastName(code);
+                    if (!String.IsNullOrEmpty(name))
+                    {
+                        try
+                        {
+                            // Attempt to get type using naming
+                            Type type = expand ? TryGetTypeFromFullName(name) : TryGetType(name);
+
+                            // CLR type
+                            if (type != null)
+                            {
+                                items = EnumerateMembers(type, name).Select(x => new PythonCodeCompletionDataCore(x.Item1, x.Item2, x.Item3, x.Item4, this));
+                            }
+                            // Variable type
+                            else if (VariableTypes.TryGetValue(name, out type))
```

Comment:
can you explain what the difference is between case 2 and case 3 here?

Reply:
`else if (VariableTypes.TryGetValue(name, out type))`
Handles the case of variable assignment (ex. a = [] then a.)
` else { var mem = LookupMember(name);`
Handles the case of Python Objects directly from the Scope (ex. modules: import sys then sys. )